### PR TITLE
Add non ICU time_bucket function

### PR DESCRIPTION
**Pull request by papparapa**
[#5589](https://github.com/duckdb/duckdb/issues/5589)

Split from PR [#5665](https://github.com/duckdb/duckdb/pull/5665).

While ICU `time_bucket` implementation in [#5665](https://github.com/duckdb/duckdb/pull/5665) still has some problems, non-ICU `time_bucket` seems ready.
This PR adds only non-ICU `time_bucket` functions.

They behave almost same as [TimescaleDB's time_bucket](https://docs.timescale.com/api/latest/hyperfunctions/time_bucket/).

```
TIME_BUCKET(bucket_width: INTERVAL, ts: TIMESTAMP)
TIME_BUCKET(bucket_width: INTERVAL, ts: DATE)

TIME_BUCKET(bucket_width: INTERVAL, ts: TIMESTAMP, offset: INTERVAL)
TIME_BUCKET(bucket_width: INTERVAL, ts: DATE, offset: INTERVAL)

TIME_BUCKET(bucket_width: INTERVAL, ts: TIMESTAMP, origin: TIMESTAMP)
TIME_BUCKET(bucket_width: INTERVAL, ts: DATE, origin: DATE)
```

### CODE REVIEW

**Review by hawkfish — COMMENTED** [review]
Looks pretty good - just use the library functions that check overflow.

File: src/function/scalar/date/time_bucket.cpp [hunk 1]
```diff
@@ -0,0 +1,386 @@
...
+	static inline timestamp_t WidthConvertibleToMicrosCommon(int64_t bucket_width_micros, int64_t ts_micros,
+	                                                         int64_t origin_micros) {
+		origin_micros %= bucket_width_micros;
+		if (origin_micros > 0 && ts_micros < NumericLimits<int64_t>::Minimum() + origin_micros) {
+			throw OutOfRangeException("Timestamp out of range");
+		}
+		if (origin_micros < 0 && ts_micros > NumericLimits<int64_t>::Maximum() + origin_micros) {
+			throw OutOfRangeException("Timestamp out of range");
+		}
+		ts_micros -= origin_micros;
```
[the remark this answers is not in the document]
Comment (papparapa) [reply src/function/scalar/date/time_bucket.cpp]: Updated

File: src/function/scalar/date/time_bucket.cpp [hunk 2]
```diff
@@ -0,0 +1,386 @@
...
+
+		int64_t result_micros = (ts_micros / bucket_width_micros) * bucket_width_micros;
+		if (ts_micros < 0 && ts_micros % bucket_width_micros != 0) {
+			if (result_micros < NumericLimits<int64_t>::Minimum() + bucket_width_micros) {
+				throw OutOfRangeException("Timestamp out of range");
+			}
+			result_micros -= bucket_width_micros;
```
[the remark this answers is not in the document]
Comment (papparapa) [reply src/function/scalar/date/time_bucket.cpp]: Updated

File: src/function/scalar/date/time_bucket.cpp [hunk 3]
```diff
@@ -0,0 +1,386 @@
...
+		}
+		if (origin_months < 0 && ts_months > NumericLimits<int32_t>::Maximum() + origin_months) {
+			throw NotImplementedException("Timestamp out of range");
+		}
+		ts_months -= origin_months;
+
+		int32_t result_months = (ts_months / bucket_width_months) * bucket_width_months;
+		if (ts_months < 0 && ts_months % bucket_width_months != 0) {
+			if (result_months < NumericLimits<int32_t>::Minimum() + bucket_width_months) {
+				throw OutOfRangeException("Timestamp out of range");
```
[the remark this answers is not in the document]
Comment (papparapa) [reply src/function/scalar/date/time_bucket.cpp]: Updated

**Review by hawkfish — APPROVED** [review]

**Comment by Mytherin** [comment]
Thanks!

**Pull request merged**